openedx/openedx-platform
# feat: default staff to courseware MFE if active (via jump_to)

Reviewers: I broke out two minor refactors into their own commits.

## Description

The /jump_to/ LMS endpoint is used in a number of places
to direct users to courseware. It currently only redirects to
Legacy courseware URLs, which then conditionally may
redirect to the Learning MFE.

Two issues with this:
1. Performance Impact: In most cases, going to Legacy first
   is just an extra redirect.
2. Confusion for Privileged Users: Neither course nor global
   staff are auto-redirected from the Legacy experience to the
   MFE. Thus, these priviliged users confusingly never see the
   MFE by default; they must always manually click into it.

This commit makes it so that /jump_to/ directs
users to whatever the default courseware experience is
for them. For staff of courses active in the new experience,
this will impact (at a minimum) the "View Live"
links in Studio, all links on the old and new LMS
course outline, and the "Resume" links on the course
dashboard. Learners should see no difference other than
a performance improvement when following courseware links
from the LMS.

This also adds an optional 'experience=[legacy|new]'
query param to /jump_to/, allowing us to specifically
generate Legacy courseware URLs for the
"View in Legacy Experience" tool.


## Supporting information

https://openedx.atlassian.net/browse/TNL-7796

## Testing instructions

### Setup

1. If testing on Devstack, `make dev.up.studio+lms+frontend-app-learning`
2. Have two courses; we'll call them CourseNew and CourseLegacy.
   * Both should be Split Mongo (`course-v1:`) courses.
   * Ideally both should have multiple sections, subsections, and units.
   * Specific content is unimportant.
3. In LMS Django Admin, add two Waffle flag overrides:
   * force new course into MFE
      * Waffle flag: `courseware.courseware_mfe`
      * Course id: id of CourseNew
      * Override choice: Force On
      * Enabled: checked
   * force legacy course 
      * Waffle flag: `courseware.courseware_mfe`
      * Course id: id of CourseLegacy
      * Override choice: Force Off
      * Enabled: checked
4. Have two users; we'll call them Staff and Learner
   * Staff should be course staff in both CourseNew and CourseLegacy
   * Learner should be staff in neither, but enrolled in both.

### Test

* Log in as Staff
  * Go to your course dashboard and click on CourseNew, which should bring you to the Legacy outline.
    * Click on Resume or any link in the outline.
    * You should be directed to the corresponding location in the MFE!
    * Click "View in Studio", which should bring you to the unit in Studio.
    * Click "View Live Version", which should bring you to the unit in the MFE.
    * Click "View in Legacy Experience", which should bring you to Legacy courseware.
    * Click around Legacy courseware a bit, confirming that you are able to use it without being forced into the MFE.
    * (While you're here, copy a Legacy courseware link for CourseNew and keep it somewhere; you will need this for the Learner portion)
    * Click "View in the New Experience", which should bring you back to the MFE.
    * Return to your course dashboard.
    * Click "Resume" on CourseNew, which should bring you to the MFE.
  * Go to your course dashboard and click on CourseLegacy, which should bring you to the legacy outline.
    * Click on Resume or any link in the outline.
    * You should be directed to the corresponding location in the Legacy experience.
    * Click "View in Studio", which should bring you to the unit in Studio.
    * Click "View Live Version", which should bring you to the unit in the Legacy experience.
    * Click "View in New Experience", which should bring you to the unit in the MFE.
    * Click around MFE courseware a bit, confirming that you are able to use it without being kicked back to Legacy.
    * (While you're here, copy an MFE courseware link for CourseLegacy and keep it somewhere; you will need this for the Learner portion)
    * Click "View in the Legacy Experience", which should bring you back to Legacy.
    * Return to your course dashboard.
    * Click "Resume" on CourseLegacy, which should drop you where you left off in Legacy courseware.
* Log in as Learner.
  * Go to your course dashboard and click on CourseNew, which should bring you to the legacy outline.
    * Click on Resume or any link in the outline.
    * You should be directed to the corresponding location in the MFE!
    * Click around a bit to confirm that you stay in the MFE.
    * Return to your course dashboard.
    * Click "Resume" on CourseNew, which should bring you to the MFE again.
    * Paste a Legacy courseware link for CourseNew into the URL bar. You should be immediately kicked back into the MFE.
  * Go to your course dashboard and click on CourseLegacy, which should bring you to the legacy outline.
    * Click on Resume or any link in the outline.
    * You should be directed to the corresponding location in the Legacy experience.
    * Click around a bit to confirm that you stay in Legacy.
    * Return to your course dashboard.
    * Click "Resume" on CourseLegacy, which should bring you to Legacy again.
    * Paste an MFE courseware link for CourseLegacy into the URL bar. You should be immediately kicked back into the Legacy experience.
   
## Deadline

Goal is to merge Monday.

## Other information

Prereqs:
- [x] https://github.com/edx/frontend-app-learning/pull/395
- [ ] https://github.com/edx/frontend-app-learning/pull/404
